lib/propolis/src/block/mod.rs

```diff
@@ -32,79 +25,126 @@ pub use in_memory::InMemoryBackend;
 mod mem_async;
 pub use mem_async::MemAsyncBackend;
 
+pub mod backend;
+pub mod device;
+
 pub type ByteOffset = usize;
 pub type ByteLen = usize;
 
 /// When `block_size` is not specified in [BackendOpts], and the backend itself
 /// is not choosing a block size, a default of 512B is used.
 pub const DEFAULT_BLOCK_SIZE: u32 = 512;
 
+#[usdt::provider(provider = "propolis")]
+mod probes {
```

Comment:
There's still the hassle of requiring both begin/complete probes if you want sizing or offset information about the operations, but I think that's an acceptable compromise, rather than having an excessive argument list for the completion probes.

An open question following the integration of this work: Should some or all of the per-device probes (nvme, virtio, etc) be stripped out, since the block-generic ones cover most of those needs?

Reply:
I do think per-device probes can be useful for correlating against in-guest operations with queue ids, command ids, etc